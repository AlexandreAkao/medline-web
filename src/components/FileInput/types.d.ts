type FileInputSize = 'small' | 'medium' | 'large';

type FileInputAccept =
  | 'application/pdf'
  | 'application/zip'
  | 'application/*'
  | 'audio/mpeg'
  | 'audio/*'
  | 'image/gif'
  | 'image/jpeg'
  | 'image/svg+xml'
  | 'image/tiff'
  | 'image/png'
  | 'image/*'
  | 'text/css'
  | 'text/html'
  | 'text/plain'
  | 'text/*'
  | 'video/mpeg'
  | 'video/*';

interface IFileInputProps {
  children: React.ReactNode;
  acceptFile?: FileInputAccept[];
  size?: FileInputSize;
  primary?: boolean;
  multiple?: boolean;
  onChange?: (files: File[]) => void;
}
