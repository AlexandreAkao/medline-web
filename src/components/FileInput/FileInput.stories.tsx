import { ComponentStory, ComponentMeta } from '@storybook/react';

import FileInput from 'components/FileInput';

export default {
  title: 'FileInput',
  component: FileInput,
} as ComponentMeta<typeof FileInput>;

const Template: ComponentStory<typeof FileInput> = args => <FileInput {...args} />;

export const AllFiles = Template.bind({});
AllFiles.args = {
  children: <div>AllFiles</div>,
};

export const OnlyImage = Template.bind({});
OnlyImage.args = {
  children: <div>OnlyImage</div>,
  acceptFile: ['image/*'],
};

export const OnlyAudio = Template.bind({});
OnlyAudio.args = {
  children: <div>OnlyAudio</div>,
  acceptFile: ['audio/*'],
};

export const OnlyVideo = Template.bind({});
OnlyVideo.args = {
  children: <div>OnlyVideo</div>,
  acceptFile: ['video/*'],
};

export const OnlyDoc = Template.bind({});
OnlyDoc.args = {
  children: <div>OnlyDoc</div>,
  acceptFile: ['application/*'],
  onChange(files) {
    console.log(files);
  },
};

export const MultipleFile = Template.bind({});
MultipleFile.args = {
  children: <div>MultipleFile</div>,
  multiple: true,
};
