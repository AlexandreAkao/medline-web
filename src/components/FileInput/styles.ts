import styled from 'styled-components';

export const FileInputContainer = styled.input`
  display: none;
`;
