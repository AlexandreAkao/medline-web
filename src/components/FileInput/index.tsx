import { ChangeEvent, useRef } from 'react';

import Button from 'components/Button';
import { FileInputContainer } from 'components/FileInput/styles';

function FileInput({ children, acceptFile, multiple, onChange, size = 'medium', primary = true }: IFileInputProps) {
  const inputRef = useRef<HTMLInputElement>(null);

  const onUploadFile = () => {
    inputRef.current?.click();
  };

  const onChangeFile = (event: ChangeEvent<HTMLInputElement>) => {
    const files = Array.from(event.target.files as FileList);

    onChange && onChange(files);
  };

  return (
    <>
      <Button size={size} primary={primary} onClick={onUploadFile}>
        {children}
      </Button>
      <FileInputContainer
        data-testid="file-input-test"
        ref={inputRef}
        type="file"
        onChange={onChangeFile}
        accept={acceptFile?.join(', ')}
        multiple={multiple}
      />
    </>
  );
}

export default FileInput;
