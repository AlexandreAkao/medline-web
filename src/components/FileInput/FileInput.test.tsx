import { render, screen } from '@testing-library/react';
import { composeStories } from '@storybook/testing-react';
import userEvent from '@testing-library/user-event';
import { vi } from 'vitest';

import * as stories from 'components/FileInput/FileInput.stories';
import FileInput from 'components/FileInput';

const { AllFiles, MultipleFile, OnlyImage, OnlyAudio, OnlyVideo, OnlyDoc } = composeStories(stories);

describe('FileInput', () => {
  const renderComponent = (customProps: Partial<IFileInputProps> = {}) => {
    const props: IFileInputProps = {
      children: 'FileInput example',
      ...customProps,
    };
    return render(<FileInput {...props} />);
  };

  const makeFakeFile = (fileName = 'image', type: FileInputAccept = 'image/jpeg'): File =>
    new File([fileName], `${fileName}.${type.split('/')[1]}`, { type });

  it('should match with snapshot', () => {
    const { container } = renderComponent();
    expect(container).toMatchSnapshot();
  });

  it('should render a input file that accept all file types', async () => {
    render(<AllFiles />);
    const buttonElement = screen.getByText(/AllFiles/i);
    const inputFileElement = screen.getByTestId('file-input-test') as HTMLInputElement;
    const files = [makeFakeFile(), makeFakeFile(), makeFakeFile()];

    await userEvent.upload(inputFileElement, files);

    expect(buttonElement).not.toBeNull();
    expect(inputFileElement.files).toHaveLength(1);
    expect(Array.from(inputFileElement.files as FileList)).toEqual([files[0]]);
  });

  it('should render a files input that accept multiple files', async () => {
    render(<MultipleFile />);
    const buttonElement = screen.getByText(/MultipleFile/i);
    const inputFileElement = screen.getByTestId('file-input-test') as HTMLInputElement;
    const files = [makeFakeFile(), makeFakeFile(), makeFakeFile()];

    await userEvent.upload(inputFileElement, files);

    expect(buttonElement).not.toBeNull();
    expect(inputFileElement.files).toHaveLength(3);
    expect(Array.from(inputFileElement.files as FileList)).toEqual(files);
  });

  it('should render a files input that accept only image', async () => {
    render(<OnlyImage multiple />);
    const buttonElement = screen.getByText(/OnlyImage/i);
    const inputFileElement = screen.getByTestId('file-input-test') as HTMLInputElement;
    const filesNotAccepted = [
      makeFakeFile('audio', 'audio/mpeg'),
      makeFakeFile('video', 'video/mpeg'),
      makeFakeFile('doc', 'application/pdf'),
    ];
    const filesAccepted = [makeFakeFile(), makeFakeFile(), makeFakeFile()];

    await userEvent.upload(inputFileElement, filesNotAccepted);
    expect(buttonElement).not.toBeNull();
    expect(inputFileElement.files).toHaveLength(0);
    expect(Array.from(inputFileElement.files as FileList)).toEqual([]);

    await userEvent.upload(inputFileElement, filesAccepted);
    expect(inputFileElement.files).toHaveLength(3);
    expect(Array.from(inputFileElement.files as FileList)).toEqual(filesAccepted);
  });

  it('should render a files input that accept only audio', async () => {
    render(<OnlyAudio multiple />);
    const buttonElement = screen.getByText(/OnlyAudio/i);
    const inputFileElement = screen.getByTestId('file-input-test') as HTMLInputElement;
    const filesNotAccepted = [
      makeFakeFile(),
      makeFakeFile('video', 'video/mpeg'),
      makeFakeFile('doc', 'application/pdf'),
    ];
    const filesAccepted = [
      makeFakeFile('audio', 'audio/mpeg'),
      makeFakeFile('audio', 'audio/mpeg'),
      makeFakeFile('audio', 'audio/mpeg'),
    ];

    await userEvent.upload(inputFileElement, filesNotAccepted);
    expect(buttonElement).not.toBeNull();
    expect(inputFileElement.files).toHaveLength(0);
    expect(Array.from(inputFileElement.files as FileList)).toEqual([]);

    await userEvent.upload(inputFileElement, filesAccepted);
    expect(inputFileElement.files).toHaveLength(3);
    expect(Array.from(inputFileElement.files as FileList)).toEqual(filesAccepted);
  });

  it('should render a files input that accept only video', async () => {
    render(<OnlyVideo multiple />);
    const buttonElement = screen.getByText(/OnlyVideo/i);
    const inputFileElement = screen.getByTestId('file-input-test') as HTMLInputElement;
    const filesNotAccepted = [
      makeFakeFile(),
      makeFakeFile('audio', 'audio/mpeg'),
      makeFakeFile('doc', 'application/pdf'),
    ];
    const filesAccepted = [
      makeFakeFile('video', 'video/mpeg'),
      makeFakeFile('video', 'video/mpeg'),
      makeFakeFile('video', 'video/mpeg'),
    ];

    await userEvent.upload(inputFileElement, filesNotAccepted);
    expect(buttonElement).not.toBeNull();
    expect(inputFileElement.files).toHaveLength(0);
    expect(Array.from(inputFileElement.files as FileList)).toEqual([]);

    await userEvent.upload(inputFileElement, filesAccepted);
    expect(inputFileElement.files).toHaveLength(3);
    expect(Array.from(inputFileElement.files as FileList)).toEqual(filesAccepted);
  });

  it('should render a files input that accept only document', async () => {
    render(<OnlyDoc multiple />);
    const buttonElement = screen.getByText(/OnlyDoc/i);
    const inputFileElement = screen.getByTestId('file-input-test') as HTMLInputElement;
    const filesNotAccepted = [makeFakeFile(), makeFakeFile('audio', 'audio/mpeg'), makeFakeFile('video', 'video/mpeg')];
    const filesAccepted = [
      makeFakeFile('doc', 'application/*'),
      makeFakeFile('doc', 'application/*'),
      makeFakeFile('doc', 'application/*'),
    ];

    await userEvent.upload(inputFileElement, filesNotAccepted);
    expect(buttonElement).not.toBeNull();
    expect(inputFileElement.files).toHaveLength(0);
    expect(Array.from(inputFileElement.files as FileList)).toEqual([]);

    await userEvent.upload(inputFileElement, filesAccepted);
    expect(inputFileElement.files).toHaveLength(3);
    expect(Array.from(inputFileElement.files as FileList)).toEqual(filesAccepted);
  });

  it('should render call ', async () => {
    const onChange = vi.fn();
    renderComponent({ onChange });
    const buttonElement = screen.getByText(/FileInput example/i);
    const inputFileElement = screen.getByTestId('file-input-test') as HTMLInputElement;
    const file = makeFakeFile();

    await userEvent.upload(inputFileElement, file);
    expect(buttonElement).not.toBeNull();
    expect(inputFileElement.files).toHaveLength(1);
    expect(Array.from(inputFileElement.files as FileList)).toEqual([file]);
    expect(onChange).toBeCalledWith([file]);
  });
});
