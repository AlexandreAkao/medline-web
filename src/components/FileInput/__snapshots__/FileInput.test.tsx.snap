// Vitest Snapshot v1

exports[`FileInput > should match with snapshot 1`] = `
<div>
  <button
    class="sc-bczRLJ eeLoKZ"
  >
    FileInput example
  </button>
  <input
    class="sc-gsnTZi iwTHrt"
    data-testid="file-input-test"
    type="file"
  />
</div>
`;
